source/exe/main_common.cc

```diff
@@ -125,10 +125,10 @@ void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string
   });
 }
 
-MainCommon::MainCommon(int argc, const char* const* argv)
+MainCommon::MainCommon(int argc, const char* const* argv, PlatformMain& platform_main)
     : options_(argc, argv, &MainCommon::hotRestartVersion, spdlog::level::info),
       base_(options_, real_time_system_, default_test_hooks_, prod_component_factory_,
-            std::make_unique<Runtime::RandomGeneratorImpl>(), thread_factory_) {}
+            std::make_unique<Runtime::RandomGeneratorImpl>(), platform_main.threadFactory()) {}
```

Comment:
Rather than changing the API to MainCommon why don't you just instantiate PlatformMain as a member var in MainCommon, at least for now. I think that would touch less code. Also this will avoid breaking our import :)

Can we rename PlatformMain to PlatformApi as there's nothing really 'Main' about it,  And actually it's taking on an overlapping role to Api::Api, bringing us back to to @htuch's https://github.com/envoyproxy/envoy/issues/5126 .

We should also either fix Api::createApiForTest in test/test_common/utility.cc to use PlatformApi, or put a TODO to do so.

Reply:
The initial idea for this PlatformSetup / PlatformMain / PlatformApi object was:
1. constructor does platform specific initialization
2. member variables / accessors for any platform specific objects (currently ThreadFactory, we'll probably add some sort of filesystem object soon)
3. destructor does platform specific cleanup

As currently implemented, we don't want the destructor (and hence WSACleanup) to be called until we're completely done.  So constructing one to use in Api::createApiForTest probably won't work (at least on Windows).

Looking at this discussion (as well as the discussion in #5126) we may want to eventually pull out 2 into a different object from 1+3. (separate the platform api from the setup). That refactor feels a little large for this specific PR, though

Reply:
I think this fits into createApiForTest but may require some auditing. The caller of that should ensure the returned Api::ApiPtr lives as long as needed.

Or is it a problem of some sort to have the ctor/dtor called on every unit test within a single process?

Reply:
Also one thing I've thought of in the past is that all Envoy tests that aren't purely portable code, could inherit from Envoy::Api::TestBase which instantiates the platform-specific API, providing access to thread systems, etc.

Reply:
We also noticed that if we created a PlatformMain inside of createApiForTest and used that to give the Api a ThreadFactory reference, the reference would become invalid once the PlatformMain is destructed. So we would have to make sure the PlatformMain lasts as long as the Api object

Reply:
Yes, that's why I think it would be easier to either have envoy tests that require such objects inherit from a new Envoy::Api::TestBase, or they can simply old the ApiPtr in a test member variable. It obviously won't work to just do `createApiForTest().threadSystem()`, and this is why I suggested to make it a TODO.

Right now there's a code-smell of having the same ifdef in two places, and I don't think you need to fix it, but we should flag it for fixing.

Reply:
Makes sense, we will add the TODO